README.md
[![Build Status](https://travis-ci.org/julienc91/react-countdown360.svg?branch=master)](https://travis-ci.org/julienc91/react-countdown360)
[![Coverage Status](https://coveralls.io/repos/github/julienc91/react-countdown360/badge.svg?branch=master)](https://coveralls.io/github/julienc91/react-countdown360?branch=master)

# react-countdown360

A simple and customizable React circular countdown that counts down a number of seconds.
This is a React implementation of John Schult's original [`jquery.countdown360`](https://github.com/johnschult/jquery.countdown360)

## Examples

You can find some examples in the [`examples` directory](https://github.com/julienc91/react-countdown360/tree/master/examples).

![Basic example][example1]
![Stylized example][example2]
![Formatter example][example3]

[example1]: https://raw.githubusercontent.com/julienc91/react-countdown360/master/doc/01_basic_countdown.gif "Baic example"
[example2]: https://raw.githubusercontent.com/julienc91/react-countdown360/master/doc/02_stylized_countdown.gif "Stylized example"
[example3]: https://raw.githubusercontent.com/julienc91/react-countdown360/master/doc/05_custom_formatters_countdown.gif "Formatter example"

## Quickstart

Install with `npm`:

```
npm install react-countdown360
```

Render your countdown:

```js
import Countdown360 from 'react-countdown360'

const App = () => {
  return <Countdown360 seconds={10} />
}
```

## Documentation

### Props

| Name                | Type     | Default                | Description                                                    |
| ------------------- | -------- | ---------------------- | -------------------------------------------------------------- |
| `seconds`           | Number   | -                      | Number of seconds for the countdown                            |
| `autoStart`         | Boolean  | `true`                 | Start the countdown immediatly after rendering                 |
| `backgroundColor`   | String   | `'#fff'`               | Color for the center of the circle                             |
| `borderFillColor`   | String   | `'#f11'`               | Color for the filled part of the border                        |
| `borderUnfillColor` | String   | `'#e6e2e7'`            | Color for the unfilled part of the border                      |
| `borderWidth`       | Number   | 20                     | Width in pixels of the border                                  |
| `clockwise`         | Boolean  | `false`                | Select the direction of rotation you prefer                    |
| `fontColor`         | String   | `'#111'`               | Font color for the label                                       |
| `fontFamily`        | String   | `'sans-serif'`         | The font to use for the label                                  |
| `fontSize`          | Number   | 45                     | Font size in pixels                                            |
| `fontWeight`        | Number   | 700                    | Font weight for the label                                      |
| `onComplete`        | Function | `undefined`            | A callback called when the countdown is over                   |
| `smooth`            | Boolean  | `false`                | Update the border once every second or smoothly                |
| `startingAngle`     | Number   | 0                      | The angle at which the countdown should start                  |
| `startingSecond`    | Number   | `undefined`            | The second at which the countdown should start                 |
| `timeFormatter`     | Func     | `timeFormatterSeconds` | A function that returns the value to display                   |
| `unitFormatter`     | Func     | `unitFormatterSeconds` | A function that returns the unit to display                    |
| `width`             | Number   | 200                    | Width in pixels of the countdown to render                     |

#### Time Formatters

You can customize the way the value on the countdown is shown by providing a custom `timeFormatter` function.

By default, the value shown is the number of seconds remaining rounded to the nearest integer.

A time formatter is a function that takes one single argument, being the number of milliseconds left, and returns
the value to show on the countdown.

For instance, the following function is a time formatter that always shows a value with at least two digits:

```js
const timeFormatterTwoDigits = timeLeft => {
  return Math.round(timeLeft / 1000).toString().padStart(2, "0")
}
```

We already provide a few time formatters that you can import from the package.

| Name                        | Description                                   | Examples                   | Suggested unit formatter |
| --------------------------- | --------------------------------------------- | -------------------------- | ------------------------ |
| `timeFormatterSeconds`      | Rounded number of seconds (default behaviour) | `0`, `1`, `12`             | `unitFormatterSeconds`   |
| `timeFormatterDigitalClock` | `MM:SS`                                       | `00:00`, `00:12`, `01: 59` | `unitFormatterBlank`     |

#### Unit Formatters

You can also customize the unit shown on the countdown by providing a custom `unitFormatter` function.

A unit formatter is a function that takes one single argument, being the value shown on the countdown (as
returned by the given time formatter), and returns the unit to display.

For instance, the following function is a unit formatter to show the number of seconds in Spanish:

```js
const unitFormatterSpanishSeconds = value => {
  return value.toString() === "1" ? "segundo" : "segundos"
}
```

Be careful when choosing your unit formatter that it matches the time formatter in use!

We already provide a few unit formatters that you can import from the package.

| Name                   | Description           | Examples            |
| ---------------------- | --------------------- | ------------------- |
| `unitFormatterSeconds` | 'second' or 'seconds' | `second`, `seconds` |
| `unitFormatterBlank`   | An empty string       |                     |

### Methods

- `start`: start or resume the countdown
- `stop`: pause the countdown
- `addSeconds (Number)`: add or remove (if negative) the given number of seconds to remaining number of seconds
- `extendTimer (Number)`: add or remove (if negative) the given number of seconds to the total number of seconds


CHANGELOG.md
# Changelog

## 1.3.2 - 12/04/2020

Bumped dependencies


## 1.3.1 - 11/01/2010

New:

* possibility to start the countdown at a specific moment with the new props `startingSecond`


## 1.3.0 - 15/10/2019

New:

* possibility to customize the angle at which the countdown starts with the new props `startingAngle`


## 1.2.0 - 14/10/2019

New:

* you can now select the direction of rotation by using the new props `clockwise`


## 1.1.1 and 1.1.2 - 11/10/2019

Fixes:

* removed some unused files from the published package


## 1.1.0 - 10/10/2019

Customize the value and unit shown on the countdown with the new props `timeFormatter`
and `unitFormatter`.

New:

* customize the value and unit shown with the new props `timeFormatter` and `unitFormatter`

Breaking:

* the `unit` props has been removed and should be replaced with the new `unitFormatter` functionality


## 1.0.0 - 07/10/2019

Complete refactoring. The `canvas` solution used by `jquery.countdown360` has been replaced with a full-CSS
implementation. Many fixes for bugs that were _imported_ from the original repo.

New:

* it is now possible to pause and resume
* you can customize the border color with the props `borderFillColor` and `borderUnfillColor`

Fixes:

* fixed issues when starting and stopping multiple times and too quickly
* fixed implementations of `start` and `stop` that would break if the countdown was already started/stopped
* fixed implementations of `extendTimer` and `addSeconds` that would break with some values

Breaking:

* some props were renamed:
  * `autostart` &rarr; `autoStart`
  * `fillStyle` &rarr; `backgroundColor`
  * `label` &rarr; `unit`
  * `strokeStyle` &rarr; `borderFillColor`
  * `strokeWidth` &rarr; `borderWidth`
* the `radius` parameter was removed in favor of a new `width` parameter
* the `startOverAfterAdding` parameter was removed


## 0.1.0 - 03/10/2019 (unreleased)

First version.
This is basically a copy and paste of the original `jquery.countdown360` plugin with React in place of jquery.


src/countdown360.js
import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { timeFormatterSeconds } from './timeFormatters'
import { unitFormatterSeconds } from './unitFormatters'

class Countdown360 extends React.Component {
  constructor (props) {
    super(props)
    const seconds = Math.max(props.seconds, 0)
    const secondsLeft =
      props.startingSecond && props.startingSecond > 0
        ? Math.min(seconds, props.startingSecond)
        : seconds

    this.state = {
      totalSeconds: seconds,
      secondsLeft: secondsLeft * 1000,
      started: false,
      lastTick: null
    }

    this.interval = null
    this.handleTick = this.handleTick.bind(this)
  }

  componentDidMount () {
    const { autoStart } = this.props
    if (autoStart) {
      this.start()
    }
  }

  componentWillUnmount () {
    clearInterval(this.interval)
    this.interval = null
  }

  componentDidUpdate (prevProps) {
    const { smooth } = this.props
    if (prevProps.smooth !== smooth) {
      clearInterval(this.interval)
      this.interval = setInterval(this.handleTick, smooth ? 16 : 1000)
    }
  }

  extendTimer (value) {
    let { secondsLeft, totalSeconds } = this.state
    totalSeconds += value
    totalSeconds = Math.max(totalSeconds, 0)

    secondsLeft = Math.min(secondsLeft, totalSeconds * 1000)
    this.setState({ secondsLeft, totalSeconds })
  }

  addSeconds (value) {
    const { secondsLeft, totalSeconds } = this.state
    value *= 1000
    this.setState({
      secondsLeft: Math.max(
        Math.min(secondsLeft + value, totalSeconds * 1000),
        0
      )
    })
  }

  start () {
    const { secondsLeft, started } = this.state
    if (started || secondsLeft <= 0) {
      return
    }

    const { smooth } = this.props
    const timerInterval = smooth ? 16 : 1000
    this.handleTick()
    this.interval = setInterval(this.handleTick, timerInterval)
    this.setState({ started: true })
  }

  stop () {
    const { started } = this.state
    if (!started) {
      return
    }
    clearInterval(this.interval)
    this.interval = null

    this.handleTick()
    this.setState({ started: false, lastTick: null })
  }

  handleTick () {
    const { secondsLeft, started } = this.state
    if (!started || secondsLeft <= 0) {
      return
    }

    const now = new Date()
    const lastTick = this.state.lastTick || now

    const diff = Math.abs(now - lastTick)
    const updatedSecondsLeft = Math.max(secondsLeft - diff, 0)
    this.setState({ lastTick: now, secondsLeft: updatedSecondsLeft }, () => {
      if (updatedSecondsLeft <= 0) {
        const { onComplete } = this.props
        onComplete && onComplete()
        this.stop()
      }
    })
  }

  render () {
    const { clockwise, smooth, startingAngle, width } = this.props
    const {
      backgroundColor,
      borderUnfillColor,
      borderFillColor,
      borderWidth
    } = this.props
    const {
      fontColor,
      fontFamily,
      fontSize,
      fontWeight,
      timeFormatter,
      unitFormatter
    } = this.props

    const { secondsLeft, totalSeconds } = this.state

    let angle
    if (!smooth) {
      angle = (Math.round(secondsLeft / 1000) / totalSeconds) * 360
    } else {
      angle = (secondsLeft / (totalSeconds * 1000)) * 360
    }

    const factor = clockwise ? -1 : 1
    const x = (angle >= 180 ? 90 - (360 - angle) : 90) * factor + startingAngle
    const y = (angle >= 180 ? 90 : -90 + angle) * factor + startingAngle

    const Wrapper = styled.div`
      align-items: center;
      background-image: linear-gradient(
          ${x}deg,
          ${angle >= 180 ? borderFillColor : borderUnfillColor} 50%,
          transparent 50%
        ),
        linear-gradient(
          ${y}deg,
          ${borderUnfillColor} 50%,
          ${borderFillColor} 50%
        );
      border-radius: 50%;
      display: flex;
      height: ${width}px;
      justify-content: center;
      width: ${width}px;
    `

    const Label = styled.div`
      align-items: center;
      background: ${backgroundColor};
      border-radius: 50%;
      box-sizing: border-box;
      color: ${fontColor};
      display: flex;
      flex-flow: column;
      font-family: ${fontFamily};
      font-size: ${fontSize}px;
      font-weight: ${fontWeight};
      height: ${width - 2 * borderWidth}px;
      justify-content: center;
      text-align: center;
      text-transform: uppercase;
      width: ${width - 2 * borderWidth}px;
    `

    const Unit = styled.div`
      text-transform: uppercase;
      font-size: ${fontSize / 3}px;
    `

    const value = timeFormatter(secondsLeft)
    const unit = unitFormatter(value)

    return (
      <Wrapper>
        <Label>
          {value}
          <br />
          <Unit>{unit}</Unit>
        </Label>
      </Wrapper>
    )
  }
}

Countdown360.defaultProps = {
  autoStart: true,
  backgroundColor: '#fff',
  borderFillColor: '#f11',
  borderUnfillColor: '#e6e2e7',
  borderWidth: 20,
  clockwise: false,
  fontColor: '#111',
  fontFamily: 'sans-serif',
  fontSize: 45,
  fontWeight: 700,
  smooth: false,
  startingAngle: 0,
  timeFormatter: timeFormatterSeconds,
  unitFormatter: unitFormatterSeconds,
  width: 200
}

Countdown360.propTypes = {
  autoStart: PropTypes.bool,
  backgroundColor: PropTypes.string,
  borderFillColor: PropTypes.string,
  borderUnfillColor: PropTypes.string,
  borderWidth: PropTypes.number,
  clockwise: PropTypes.bool,
  fontColor: PropTypes.string,
  fontFamily: PropTypes.string,
  fontSize: PropTypes.number,
  fontWeight: PropTypes.number,
  onComplete: PropTypes.func,
  seconds: PropTypes.number.isRequired,
  smooth: PropTypes.bool,
  startingAngle: PropTypes.number,
  startingSecond: PropTypes.number,
  timeFormatter: PropTypes.func,
  unitFormatter: PropTypes.func,
  width: PropTypes.number
}

export default Countdown360
